File: python/aqc_apply.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

#
# @author Vladimir S. FONOV
# @date 13/04/2018
import argparse
import os
import sys

import numpy as np
import io
import copy

from aqc_data import *
from model.util import *

import torch
import torch.nn as nn

from torch.autograd import Variable


default_data_dir=os.path.dirname(sys.argv[0])
if default_data_dir=='' or default_data_dir is None: default_data_dir='.'

def parse_options():

    parser = argparse.ArgumentParser(description='Apply automated QC',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument("--ref", action="store_true", default=False,
                        help="Use reference images")
    parser.add_argument("--image", type=str, 
                        help="Input image prefix: <prefix>_{0,1,2}.jpg")
    parser.add_argument("--volume", type=str, 
                        help="Input minc volume (need minc2 simple)")
    parser.add_argument("--resample", type=str, 
                        help="Resample to standard space using provided xfm, needs mincresample")
    parser.add_argument("--low", type=float, 
                        help="Winsorize intensities in the input image", default=5.0)
    parser.add_argument("--high", type=float, 
                        help="Winsorize intensities in the input image", default=95.0)
    parser.add_argument("--load", type=str, default=None,
                        help="Load pretrained model (will try to  load from {}".format(default_data_dir))
    parser.add_argument("--net", choices=['r18', 'r34', 'r50','r101','r152','sq101'],
                        help="Network type",default='r18')
    parser.add_argument('--raw', action="store_true", default=False,
                        help='Print raw score [0:1]')
    parser.add_argument('-q', '--quiet', action="store_true",default=False,   
                        help='Quiet mode, set status code to 0 - Pass, 1 - fail')
    parser.add_argument('--batch', type=str,   
                        help='Process minc files in batch mode, provide list of files')
    parser.add_argument('--batch-size', type=int, default=1, 
                        dest='batch_size',
                        help='Batch size in batch mode')
    parser.add_argument('--batch-workers', type=int, default=1,
                        dest='batch_workers',
                        help='Number of workers in batch mode')
    parser.add_argument('--batch_pics', action="store_true", default=False,
                        help='Process QC pics in batch mode instead of MINC volumes')
    parser.add_argument('--gpu', action="store_true", default=False,
                        help='Run inference in gpu')
    parser.add_argument("--dist",action="store_true",default=False,
                        help="Predict misregistration distance instead of class membership")

    params = parser.parse_args()
    
    return params


if __name__ == '__main__':
    params = parse_options()
    use_ref = params.ref

    if params.load is None:
        params.load = default_data_dir + os.sep \
            + 'model_' + params.net + ('_ref' if params.ref else '') + os.sep + \
            'best_tnr.pth' # if params.gpu else 'best_tnr_cpu.pth' )
    
    model = get_qc_model(params, use_ref=use_ref, predict_dist=params.dist)
    model.train(False)

    if params.gpu:
        model=model.cuda()


    with torch.no_grad():
        if params.batch is not None:
            if not params.batch_pics :
                dataset = MincVolumesDataset(csv_file=params.batch,
                    winsorize_low=params.low,
                    winsorize_high=params.high,
                    data_prefix=default_data_dir + "../data",
                    use_ref=use_ref) 
            else:
                dataset = QCImagesDataset(csv_file=params.batch,
                            data_prefix=default_data_dir + "/../data",
                            use_ref=use_ref)
            
            dataloader = DataLoader(dataset, 
                            batch_size=params.batch_size,
                            shuffle=False, 
                            num_workers=params.batch_workers)

            for i_batch, sample_batched in enumerate(dataloader):
                inputs, files = sample_batched
                if params.gpu: inputs = inputs.cuda()
                outputs = model(inputs)
                if params.gpu: outputs = outputs.cpu()
                
                if params.dist:
                    if params.raw:
                        preds   = outputs[:,0].tolist()
                    else:
                        preds   = torch.max(outputs, 1)[1].tolist()

                    for i,j in zip(files,preds):
                        print("{},{}".format(i,j))
                else:
                    outputs = nn.functional.softmax(outputs,1)
                    if params.raw:
                        preds   = outputs[:,1].tolist()
                    else:
                        preds   = torch.max(outputs, 1)[1].tolist()

                    for i,j in zip(files,preds):
                        print("{},{}".format(i,j))
        else:
            if params.image is not None:
                inputs = load_qc_images( [params.image+'_0.jpg', params.image+'_1.jpg', params.image+'_2.jpg'])
            elif params.volume is not None:
                tmpdir=None
                volume=params.volume
                if params.resample is not None:
                    import tempfile,subprocess,shutil
                    tmpdir=tempfile.mkdtemp(prefix='deep_qc')
                    tmp_vol=tmpdir+os.sep+'tmp.mnc'
                    # provide sampling in the standard space
                    try:
                        args=['mincresample', '-q' ,'-transform',params.resample,
                            '-dircos', '1' ,'0', '0','0', '1', '0', '0', '0', '1', 
                            '-step', '1', '1', '1', 
                            '-start', '-96', '-132', '-78', 
                            '-nelements', '193', '229', '193', params.volume, tmp_vol,'-tfm_input_sampling']
                        subprocess.check_call(args)
                    except:
                        shutil.rmtree(tmpdir)
                        raise
                    volume=tmp_vol
                inputs = load_minc_images(volume,winsorize_low=params.low,winsorize_high=params.high)
                if tmpdir is not None:
                    shutil.rmtree(tmpdir)
            else:
                print("Specify input volume or image prefix or batch list")
                exit(1)


            # convert inputs into properly formated tensor
            # with a single batch dimension
            inputs = torch.cat( inputs ).unsqueeze_(0)

            outputs = model(inputs)

            if params.dist:
                outputs = outputs[0,0]
                preds   = (outputs>10.0).squeeze() # TODO : parametrize threshold
            else:
                outputs = nn.functional.softmax(outputs,1)
                preds   = torch.max(outputs,1)[1].squeeze()
                outputs = outputs[0,1]
            # raw score
            if params.raw:
                print(float(outputs))
            elif not params.quiet:
                if preds:
                    print("Pass")
                else:
                    print("Fail")
            else:
                exit(0 if preds[0]==1 else 1)

File: lua/aqc_apply.lua
#! /usr/bin/env th

require 'torch'
require 'xlua'
require 'paths'
require 'nn'
require 'image'
-- require 'minc2_simple'

torch.setdefaulttensortype('torch.FloatTensor')

function load_ref_data( img )
    local i,j
    local ret={}
    
    for i,j in ipairs(img) do
        ret[i]=image.load( j, 1, 'float' )
    end
    
    return ret
end

function parse(arg)
   local cmd = torch.CmdLine()
   local scriptdir=paths.dirname(paths.thisfile(arg))..'/..'
   
   cmd:text()
   cmd:text('Torch-based Auto QC application script ')
   cmd:text()
   cmd:text('Options:')
    ------------ General options --------------------
   cmd:option('-data',  scriptdir, 'Data prefix')
   cmd:option('-model', '',     'Alternative pre-trained model')
   cmd:option('-noref', false,  "Don't use reference")
   cmd:option('-gpu',   false,  'Use GPU/cudnn (model should be for GPU)')
   cmd:option('-volume', '',    'Input volume (minc)')
   cmd:option('-image', '',     'Input image base, will search for <base>_0.jpg, <base>_1.jpg, <base>_2.jpg ')
   cmd:option('-raw',  false,   'Print raw score [0:1]')
   cmd:option('-q',    false,   'Quiet mode, set status code to 0 - Pass, 1 - fail')
   cmd:text()
   local opt = cmd:parse(arg or {})
   
   return opt
end

local opt=parse(arg)

if not (opt.volume ~= "" or opt.image ~= "") then
    print(string.format("Need to specify input volume or image"))
    return 1
end

if opt.gpu then
    require 'cutorch'
    require 'cunn'
    require 'cudnn'
end

local model=nil
if opt.model ~= '' then
    model=torch.load(opt.model)
else
    model=torch.load(opt.data..'/results/r18_ref_01_training_cpu.t7')
end

local mean_sd=torch.load(opt.data..'/results/r18_ref_mean_sd.t7')

model:evaluate()

if not opt.noref then
qc_ref=load_ref_data({opt.data.."/data/mni_icbm152_t1_tal_nlin_sym_09c_0.jpg",
                      opt.data.."/data/mni_icbm152_t1_tal_nlin_sym_09c_1.jpg",
                      opt.data.."/data/mni_icbm152_t1_tal_nlin_sym_09c_2.jpg"})
end

if opt.image ~= "" then -- load JPGs from file
    input_image=load_ref_data({opt.image.."_0.jpg",opt.image.."_1.jpg",opt.image.."_2.jpg"} )
else
    input_minc=minc2_file.new(opt.volume)
    input_minc:setup_standard_order()

    -- load into standard volume
    local sample=input_minc:load_complete_volume(minc2_file.MINC2_FLOAT)
    
    -- normalize input
    local min=sample:min()
    local max=sample:max()
    sample=sample:add(-min):mul(1.0/(max-min))
    
    local sz=sample:size()
    
    input_image={sample[{sz[1]/2+1,{},{}}],
                 sample[{{},{},sz[3]/2+1}],
                 sample[{{},sz[2]/2+1,{}}]
                 }
    
    -- flip, resize and crop
    for i=1,3 do
        input_image[i]=image.scale(image.vflip(input_image[i]:contiguous()),256)
        sz=input_image[i]:size()
        
        --pad image 
        local dummy=torch.Tensor(256,256):zero()
        
        dummy[{ {(256-sz[1])/2+1, (256-sz[1])/2+sz[1]},{(256-sz[2])/2+1,(256-sz[2])/2+sz[2]}} ]=input_image[i]
        -- crop
        input_image[i]=image.crop(dummy,'c',224,224)
    end
end

if not opt.noref then
    data_in=torch.Tensor(     1, 6 , 224, 224 )
else
    data_in=torch.Tensor(     1, 3 , 224, 224 )
end

stride=1
if not opt.noref then stride=2 end

-- apply pre-scaling
for i=1,3 do
    input_image[i]:add( -mean_sd.mean  )
    input_image[i]:mul( 1.0/mean_sd.sd )
    
    
    data_in[{1,(i-1)*stride+1,{},{}}]=input_image[i]
    if not opt.noref then
        qc_ref[i]:add( -mean_sd.mean  )
        qc_ref[i]:mul( 1.0/mean_sd.sd )
        data_in[{1,(i-1)*stride+2,{},{}}]=qc_ref[i]
    end
end

-- get the result
local output = model:forward(data_in)
local _,out_max = output:max(2)
out_max=out_max:storage()[1]
out_p=output:exp():storage()[2]

if opt.raw then
    print(out_p)
elseif not opt.quiet then
    if out_max==2 then 
        print("PASS")
    else 
        print("FAIL")
    end
else
    -- return exit status
    os.exit(2-out_max)
end

-- 
